opt: set min row count of 1 for FK and uniqueness check WithScans

mgartner opened the pull request:
#### opt: add optimizer_check_input_min_row_count

Informs #137547

Release note (sql change): The `optimizer_check_input_min_row_count`
session setting has been added to control the minimum row count estimate
for buffer scans of foreign key and unqiueness checks. It defaults to 0.

#### opt: default optimizer_check_input_min_row_count to 1

Fixes #137547

Release note (performance improvement): The
`optimizer_check_input_min_row_count` session settings now defaults to
1, resulting in better query plans for foreign key and uniqueness
checks.

mgartner committed:
opt: add optimizer_check_input_min_row_count

Informs #137547

Release note (sql change): The `optimizer_check_input_min_row_count`
session setting has been added to control the minimum row count estimate
for buffer scans of foreign key and unqiueness checks. It defaults to 0.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140735)
<!-- Reviewable:end -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice!

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-:-OIXi3VKCb4pWhQ_skm2:b-8gk8uj)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
:lgmt: Should we backport this with a session var disabled by default?

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-:-OIXn5V0-wCwrCkmyY_6:b-mi5b71)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Cool!

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-:-OIZSqAB-tKWMbONUKoA:b-8mhwdz)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 1773 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIZSUQ52mXy6xFvh1ra:-OIZSUQ52mXy6xFvh1rb:bogfshi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d8f549d5bda0a5cbeebf60f4de16c25ba3342d0/pkg/sql/opt/optbuilder/mutation_builder.go#L1773)):*
> ```Go
> 		OutCols:      outScope.colList(),
> 		ID:           mb.b.factory.Metadata().NextUniqueID(),
> 		FKCheckInput: true,
> ```

nit: this also gets set for uniqueness checks, right? That's probably desirable, but makes the naming a little confusing.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Given that the same customer is running into the same issue again ([ticket](https://github.com/cockroachlabs/support/issues/3184)), and we don't have a good workaround, I do think we want to backport to at least 24.3 branch.

yuzefovich reviewed (commented):
I added a session variable to make it backportable. We can bikeshed on the naming - I had hard time coming up with something descriptive yet consice.

> Reviewed 1 of 1 files at r2, 11 of 11 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-:-OIrzBBc6ozkIUw_-bOF:b4cindb)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball)

___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 1773 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIZSUQ52mXy6xFvh1ra:-OIrvVkvA_yhwfDmIWVE:b-vz090i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d8f549d5bda0a5cbeebf60f4de16c25ba3342d0/pkg/sql/opt/optbuilder/mutation_builder.go#L1773)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: this also gets set for uniqueness checks, right? That's probably desirable, but makes the naming a little confusing.
</blockquote></details>

You're right, and I agree that it seems desirable. Updated the naming and the comments.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
I'm curious, does setting `optimizer_min_row_count` also fix this? I wonder if this is just a special case of that more general fix?

> Reviewed 2 of 11 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-:-OIsi3byEXEtpEPcuZUU:b82vobn)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball and @yuzefovich)

___
*[`-- commits` line 8 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIscO_J9OGpK_5e8Od7:-OIscO_J9OGpK_5e8Od8:bkszb40):*
we should mention the session variable in the release note
___
*[`pkg/sql/vars.go` line 3806 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIsijY64m9EQrLzOJ85:-OIsijY64m9EQrLzOJ86:b-geo1h4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e254eb1a1666cd40edfd52606837abb94e7ea27/pkg/sql/vars.go#L3806)):*
> ```Go
> 
> 	// CockroachDB extension.
> 	`opt_fk_uniq_check_lower_bound_enabled`: {
> ```

How about `optimizer_fk_min_row_count` or `optimizer_checks_min_row_count` to match `optimizer_min_row_count`?

Also, should this be a float rather than a boolean, to give ourselves more flexibility?
___
*[`pkg/sql/opt/memo/memo.go` line 204 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIshK6OB2Htn3K5CxBk:-OIshK6OB2Htn3K5CxBl:b-hld3ly) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e254eb1a1666cd40edfd52606837abb94e7ea27/pkg/sql/opt/memo/memo.go#L204)):*
> ```Go
> 	legacyVarcharTyping                        bool
> 	preferBoundedCardinality                   bool
> 	useFKUniqCheckLowerBoundOne                bool
> ```

micro-nit: would be nice to keep the same name throughout, to help our future selves


<!-- Sent from Reviewable.io -->

mgartner commented on pkg/sql/vars.go:
```diff
@@ -3801,6 +3801,23 @@ var varGen = map[string]sessionVar{
+			m.SetOptFKUniqCheckLowerBoundEnabled(b)
+			return nil
+		},
+		GlobalDefault: globalTrue,
```
Backports can be simpler if you set this to false in the first commit, and set it to true in another commit. Then you can backport just the first commit cleanly without having to update pg_catalog, show_source, and information_schema tests.

mgartner replied:
I went ahead and did this.

mgartner replied:
Of course, the backports weren't clean because I didn't do the smart thing you did with the session setting orderings. Oh well... They are here if we want them:

https://github.com/cockroachdb/cockroach/pull/141374
https://github.com/cockroachdb/cockroach/pull/141375
https://github.com/cockroachdb/cockroach/pull/141376
https://github.com/cockroachdb/cockroach/pull/141378

mgartner committed:
opt: default optimizer_check_input_min_row_count to 1

Fixes #137547

Release note (performance improvement): The
`optimizer_check_input_min_row_count` session settings now defaults to
1, resulting in better query plans for foreign key and uniqueness
checks.

mgartner reviewed (commented):
Made a few more naming changes and split into two commits - one to add the setting with it defaulted to `0`, and one to update the default to `1`. This should make the backports simple to recreate, if we'd like to.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-:-OIuqT9s-EKF79hkcOKa:b-1n348s)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball, @michae2, and @yuzefovich)

___
*[`-- commits` line 8 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIscO_J9OGpK_5e8Od7:-OIuqIbv0hErnbzddaYI:b-896fix):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

we should mention the session variable in the release note
</blockquote></details>

Done.
___
*[`pkg/sql/vars.go` line 3806 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIsijY64m9EQrLzOJ85:-OIuqK5uEFv0k6RAXlAw:b-nkwo4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e254eb1a1666cd40edfd52606837abb94e7ea27/pkg/sql/vars.go#L3806)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

How about `optimizer_fk_min_row_count` or `optimizer_checks_min_row_count` to match `optimizer_min_row_count`?

Also, should this be a float rather than a boolean, to give ourselves more flexibility?
</blockquote></details>

Made it a float.
___
*[`pkg/sql/opt/memo/memo.go` line 204 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIshK6OB2Htn3K5CxBk:-OIuqMgx4BvPrX9j0Axm:b-896gsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e254eb1a1666cd40edfd52606837abb94e7ea27/pkg/sql/opt/memo/memo.go#L204)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

micro-nit: would be nice to keep the same name throughout, to help our future selves
</blockquote></details>

Don.e
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 1773 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIZSUQ52mXy6xFvh1ra:-OIuqOUkEsygcAVotkrE:bocd1my) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d8f549d5bda0a5cbeebf60f4de16c25ba3342d0/pkg/sql/opt/optbuilder/mutation_builder.go#L1773)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

You're right, and I agree that it seems desirable. Updated the naming and the comments.
</blockquote></details>

Simplified further to `CheckInput`.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks! I just have one nit.

> Reviewed 1 of 1 files at r4, 12 of 12 files at r5, 5 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-:-OIv7lZN7vOIH307kv4l:b-t80grj)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @michae2)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 2754 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIv7UFvB2kODTz4RpLw:-OIv7UFvB2kODTz4RpLx:baebomh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3aa42fa0c64bdb6255e5cfaeb008643265a955/pkg/sql/opt/memo/statistics_builder.go#L2754)):*
> ```Go
> 	s.RowCount = bindingProps.Statistics().RowCount
> 	if withScan.CheckInput && sb.checkInputMinRowCount > 0 {
> 		// Assume that WithScans that are leaf expressions of FK and uniqueness
> ```

nit: should we remove this comment given that we're not hard-coding row count 1 anymore?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140735)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 2753 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIv8L-cDulNbgn0i7Bi:-OIv8L-cDulNbgn0i7Bj:b-mvsq9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/976cdf74f6ea4873ab87d4b0c86b8784be468324/pkg/sql/opt/memo/statistics_builder.go#L2753)):*
> ```Go
> 	s.Available = bindingProps.Statistics().Available
> 	s.RowCount = bindingProps.Statistics().RowCount
> 	if withScan.CheckInput && sb.checkInputMinRowCount > 0 {
> ```

nit: we could also remove `sb.checkInputMinRowCount > 0` condition now since max will work in zero case too.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140735)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @michae2 and @yuzefovich)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 2754 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIv7UFvB2kODTz4RpLw:-OIvKT2O1dIaE5SdiPd-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae3aa42fa0c64bdb6255e5cfaeb008643265a955/pkg/sql/opt/memo/statistics_builder.go#L2754)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we remove this comment given that we're not hard-coding row count 1 anymore?
</blockquote></details>

Done.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 2753 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-OIv8L-cDulNbgn0i7Bi:-OIvKS3GD2LQAt0NfdTl:boj1qgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/976cdf74f6ea4873ab87d4b0c86b8784be468324/pkg/sql/opt/memo/statistics_builder.go#L2753)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we could also remove `sb.checkInputMinRowCount > 0` condition now since max will work in zero case too.
</blockquote></details>

Good catch. Done.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140735#-:-OIvZqoM9jJV99ZvXUyX:b-8gk8uj)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs and assistance!

bors r+
